pkg/ddl/notifier/subscribe.go

```diff
@@ -0,0 +1,210 @@
+	store        Store
+	handlers     map[HandlerID]SchemaChangeHandler
+	pollInterval time.Duration
+}
+
+var globalDDLNotifier *ddlNotifier
+
+// InitDDLNotifier initializes the global ddlNotifier. It should be called only
+// once and before any RegisterHandler call. The ownership of the sctx is passed
+// to the ddlNotifier.
+func InitDDLNotifier(
+	sctx sessionctx.Context,
+	store Store,
+	pollInterval time.Duration,
+) {
+	globalDDLNotifier = &ddlNotifier{
+		ownedSCtx:    sctx,
+		store:        store,
+		handlers:     make(map[HandlerID]SchemaChangeHandler),
+		pollInterval: pollInterval,
+	}
+}
+
+// ResetDDLNotifier is used for testing only.
+func ResetDDLNotifier() { globalDDLNotifier = nil }
+
+// StartDDLNotifier starts the global ddlNotifier. It will block until the
+// context is canceled.
+func StartDDLNotifier(ctx context.Context) {
+	globalDDLNotifier.Start(ctx)
+}
+
+// Start starts the ddlNotifier. It will block until the context is canceled.
+func (n *ddlNotifier) Start(ctx context.Context) {
+	ctx = kv.WithInternalSourceType(ctx, kv.InternalDDLNotifier)
+	ticker := time.NewTicker(n.pollInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			if err := n.processEvents(ctx); err != nil {
+				logutil.Logger(ctx).Error("Error processing events", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (n *ddlNotifier) processEvents(ctx context.Context) error {
+	changes, err := n.store.List(ctx, sess.NewSession(n.ownedSCtx))
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// we should ensure deliver order of events to a handler, so if a handler returns
+	// error for previous events it should not receive later events.
+	skipHandlers := make(map[HandlerID]struct{})
+	for _, change := range changes {
+		for handlerID, handler := range n.handlers {
+			if _, ok := skipHandlers[handlerID]; ok {
+				continue
+			}
+			if err2 := n.processEventForHandler(ctx, change, handlerID, handler); err2 != nil {
+				skipHandlers[handlerID] = struct{}{}
+
+				if !goerr.Is(err2, ErrNotReadyRetryLater) {
+					logutil.Logger(ctx).Error("Error processing change",
+						zap.Int64("ddlJobID", change.ddlJobID),
+						zap.Int64("multiSchemaChangeSeq", change.multiSchemaChangeSeq),
+						zap.Int("handlerID", int(handlerID)),
+						zap.Error(err2))
+				}
+				continue
+			}
+		}
+		// TODO: remove the processed change after all handlers processed it.
+	}
+
+	return nil
+}
+
+func (n *ddlNotifier) processEventForHandler(
+	ctx context.Context,
+	change *schemaChange,
+	handlerID HandlerID,
+	handler SchemaChangeHandler,
+) (err error) {
+	if (change.processedByFlag & (1 << handlerID)) != 0 {
+		return nil
+	}
+
+	se := sess.NewSession(n.ownedSCtx)
+
+	if err = se.Begin(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	defer func() {
+		if err == nil {
+			err = errors.Trace(se.Commit(ctx))
+		} else {
+			se.Rollback()
+		}
+	}()
+
+	// TODO: Should we attach a timeout to this ctx?
```

Comment:
Another idea is to record the time spent and print a log when it exceeds a threshold.